src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,137 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String INDEX_OUT_OF_BOUNDS_FORMAT =
+            "Index = [%d] was out of bounds for size = [%d]";
+    private static final int LAST_ELEMENT_INDEX_SHIFT = 1;
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(value);
+        if (isEmpty()) {
+            head = node;
+        } else {
+            tail.next = node;
+            node.previous = tail;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> nextNode = getNodeByIndex(index);
+        Node<T> previousNode = nextNode.previous;
+        previousNode.next = newNode;
+        newNode.previous = previousNode;
+        newNode.next = nextNode;
+        nextNode.previous = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index == 0) {
+            T oldValue = head.value;
+            head = head.next;
+            size--;
+            return oldValue;
+        }
+        if (index >= 0 && index == size - LAST_ELEMENT_INDEX_SHIFT) {
+            T oldValue = tail.value;
+            tail = tail.previous;
+            size--;
+            return oldValue;
+        }
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        nodeToRemove.previous.next = nodeToRemove.next;
+        nodeToRemove.next.previous = nodeToRemove.previous;
+        size--;
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
```

Comment:
Don't forget that you can have `null` as a method parameter in the `remove(T object)` method. You need to handle this case.